iocore/net/quic/QUICFrame.cc

```diff
@@ -2534,312 +2405,295 @@ QUICRetireConnectionIdFrame::seq_num() const
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_stream_id_blocked_frame);
+    new (buf) QUICStreamIdBlockedFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::NEW_CONNECTION_ID:
-    frame = quicNewConnectionIdFrameAllocator.alloc();
-    new (frame) QUICNewConnectionIdFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_new_connection_id_frame);
+    new (buf) QUICNewConnectionIdFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::STOP_SENDING:
-    frame = quicStopSendingFrameAllocator.alloc();
-    new (frame) QUICStopSendingFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_stop_sending_frame);
+    new (buf) QUICStopSendingFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::PATH_CHALLENGE:
-    frame = quicPathChallengeFrameAllocator.alloc();
-    new (frame) QUICPathChallengeFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_path_challenge_frame);
+    new (buf) QUICPathChallengeFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::PATH_RESPONSE:
-    frame = quicPathResponseFrameAllocator.alloc();
-    new (frame) QUICPathResponseFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_path_response_frame);
+    new (buf) QUICPathResponseFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::NEW_TOKEN:
-    frame = quicNewTokenFrameAllocator.alloc();
-    new (frame) QUICNewTokenFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_new_token_frame);
+    new (buf) QUICNewTokenFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   case QUICFrameType::RETIRE_CONNECTION_ID:
-    frame = quicRetireConnectionIdFrameAllocator.alloc();
-    new (frame) QUICRetireConnectionIdFrame(buf, len);
-    return QUICFrameUPtr(frame, &QUICFrameDeleter::delete_retire_connection_id_frame);
+    new (buf) QUICRetireConnectionIdFrame(src, len);
+    return reinterpret_cast<QUICFrame *>(buf);
   default:
     // Unknown frame
-    Debug("quic_frame_factory", "Unknown frame type %x", buf[0]);
-    return QUICFrameFactory::create_null_frame();
+    Debug("quic_frame_factory", "Unknown frame type %x", src[0]);
+    return nullptr;
   }
 }
 
-std::shared_ptr<const QUICFrame>
+const QUICFrame &
 QUICFrameFactory::fast_create(const uint8_t *buf, size_t len)
 {
   if (QUICFrame::type(buf) == QUICFrameType::UNKNOWN) {
-    return nullptr;
+    return this->_unknown_frame;
   }
 
-  std::shared_ptr<QUICFrame> frame = this->_reusable_frames[static_cast<ptrdiff_t>(QUICFrame::type(buf))];
+  QUICFrame *frame = this->_reusable_frames[static_cast<ptrdiff_t>(QUICFrame::type(buf))];
 
   if (frame == nullptr) {
-    frame = QUICFrameFactory::create(buf, len);
+    frame = QUICFrameFactory::create(static_cast<uint8_t *>(malloc(1024)), buf, len);
```

Comment:
When this buffer will be freed?

Reply:
Currently not, but maybe in the destructor, or maybe we should use MemArena.